### DISCUSSION TITLE
SQLAlchemy 1.4 async event listeners

### DISCUSSION THREAD

**Discussion by cmflynn** [sqlalchemy/sqlalchemy#6594] [Usage Questions] [answered]
```python
engine: Engine = create_async_engine(
    URL, echo=True, future=True
)

async_session = sessionmaker(
    engine, class_=AsyncSession, expire_on_commit=False, future=True
)


event.listens_for(engine, "do_connect")(do_connect_listener)
event.listens_for(engine, "engine_connect")(engine_connect_listener)
```
Currently raises a `{NotImplementedError}asynchronous events are not implemented at this time.  Apply synchronous listeners to the AsyncEngine.sync_engine or AsyncConnection.sync_connection attributes.`.  

I'm struggling to find any other additional information on this. Will event handlers ever be supported in async sqlalchemy? If not, are there any documented alternatives to handle things like do_connect and engine_connect handles?

**Comment by CaselIT** [comment, marked as answer]
Hi,

As suggested by the exception you may use the sync_engine/sync_connection when applying events. Here is an example:
```py
import asyncio
from sqlalchemy import event
from sqlalchemy.ext.asyncio import create_async_engine

engine = create_async_engine("sqlite+aiosqlite:///:memory:")

@event.listens_for(engine.sync_engine, "do_connect")
def do_connect(dialect, conn_rec, cargs, cparams):
    print("some-function")

@event.listens_for(engine.sync_engine, "engine_connect")
def engine_connect(conn, branch):
    print("engine_connect", conn.exec_driver_sql("select 1").scalar())

async def go():
    async with engine.connect() as conn:
        res = await conn.exec_driver_sql("select 2")
        print("go", res.scalar())

asyncio.run(go())
```

> Will event handlers ever be supported in async sqlalchemy?

we have it in the to-do list, but it's not currently a top priority since using the sync version is a viable alternative

**Reply by cmflynn** [reply]
Nice! I wasn't sure how to apply the information error message like this. Thanks so much!

**Reply by Neki** [reply]
Hello,

I just encountered this use case. When authenticating to AWS RDS using IAM auth, the DB password should be retrieved by HTTP with an API call.
The simplest implementation would be to have a `do_connect` event listener that fetches and injects the password in the connection parameters.
If the event listener is not async, will it block the event loop during the HTTP call? (it seems like it does in my testing, but maybe I'm missing something).

**Reply by CaselIT** [reply]
It depends of what http library you are using. I think you can use an async http library with our `await_only` wrapper, since the event is run in the proper context.
But that's not really a public api and is not documented.

Note that the `do_connect ` is called every time a connection to the db is initiated. If the password does not change while the application is running, it may make more sense to fetch them before creating the engine the first time.

**Reply by zzzeek** [reply]
this is a little bit of a use case though isn't it.    at some point we should document, if someone is in a "run_sync" context, and you need to run an awaitable, here's how....

since we have some "dynamic access token" examples coming in already, like https://docs.sqlalchemy.org/en/14/core/engines.html#generating-dynamic-authentication-tokens and https://docs.sqlalchemy.org/en/14/dialects/mssql.html#mssql-pyodbc-access-tokens , we might want to also talk about AWS use cases including sync and async.

**Reply by CaselIT** [reply]
> this is a little bit of a use case though isn't it. at some point we should document, if someone is in a "run_sync" context, and you need to run an awaitable, here's how....

it makes sense, but I think we should have a separate public api for it. Like instead of `await_only` an `await_coroutine`, otherwise you would need to explain why there is also ` await_fallback`, when to use one and not the other, why etc. All in all that is probably best to keep as an implementation detail of sqlalchemy.
This `await_coroutine` could be an alias for `await_only` now, but we are free to make it something else in the future

**Reply by Neki** [reply]
Thanks for the replies! I'll try `await_only` in the meantime, and will move to an officially supported API if there's one in a future release.
I can use an async library to perform the HTTP calls, but I can't do it at engine creation, because the password is only valid for 60 seconds.

**Reply by CaselIT** [reply]
> because the password is only valid for 60 seconds.

wow, that seems a bit overkill.

Let us know if that works

**Reply by zzzeek** [reply]
> > this is a little bit of a use case though isn't it. at some point we should document, if someone is in a "run_sync" context, and you need to run an awaitable, here's how....
> 
> it makes sense, but I think we should have a separate public api for it. Like instead of `await_only` an `await_coroutine`, otherwise you would need to explain why there is also ` await_fallback`, when to use one and not the other, why etc. All in all that is probably best to keep as an implementation detail of sqlalchemy.
> This `await_coroutine` could be an alias for `await_only` now, but we are free to make it something else in the future

of course, we should name it something and place it in a nice place and all that.

**Reply by zzzeek** [reply]
> > because the password is only valid for 60 seconds.
> 
> wow, that seems a bit overkill.
> 
> Let us know if that works

this is common for these things.  it's an access token that is assumed to be acquired programmatically.    it is likely possible for an extension library to implement these patterns out of the box using CreateEnginePlugin.

**Comment by mark-barrett** [comment]
Hey, I have the same issue but with `SessionEvents`:
https://docs.sqlalchemy.org/en/20/orm/events.html#session-events

When I try to use `SessionLocal.sync_engine`
`SessionLocal = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)`

it says "sessionmaker has no attribute sync_session"

when I try to use `session.sync_session`:
`session = scoped_session(SessionLocal)`

it says scoped_session has no attribute sync_session

and finally, when I try to use the actual class AsyncEngine.sync_engine (I know this won't work)
it says I can't use the Class type obviously

I think I'm missing something, do you know what specific session object it refers to?

**Reply by zzzeek** [reply]
there's [async_sessionmaker](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.async_sessionmaker) now, but as yet there is not a recipe to associate SessionEvents with the async_sessionmaker itself.     

here's one

```py
import asyncio

from sqlalchemy import event
from sqlalchemy.ext.asyncio import async_sessionmaker
from sqlalchemy.orm import sessionmaker

sync_maker = sessionmaker()
maker = async_sessionmaker(sync_session_class=sync_maker)

@event.listens_for(sync_maker, "before_commit")
def before_commit(session):
    print("before commit")


async def main():
    async_session = maker()

    await async_session.commit()

asyncio.run(main())



```

**Reply by zzzeek** [reply]
I'm adding docs for this now.  Check at https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#using-events-with-the-asyncio-extension in about an hour for updates

**Reply by mark-barrett** [reply]
Oh fantastic, thanks so much, appreciate the quick reply! 🙂

**Reply by zzzeek** [reply]
here we go https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#examples-of-event-listeners-with-async-engines-sessions-sessionmakers

**Comment by PaleNeutron** [comment]
Sorry for open this old discussion, but if we do some slow query or insert in sync event environment I think it will block the whole thread.

For example I have a orm listener to create a new history object on origin object changes.

Can we have a event listener for async function?

It may looks like:

```python
@event.listens_for_async(Portfolio, "before_update")
async def create_portfolio_his(mapper, connection, target, *args, **kwargs):
    ...
````

**Reply by CaselIT** [reply]
if you run a query using the provided connection/session then that query will not block the thread. if you use other sync blocking api that aren't controlled by sqlalchemy you will block the thread.
You can run the async version of those blocking api using https://docs.sqlalchemy.org/en/20/core/internals.html#sqlalchemy.engine.AdaptedConnection.run_async

**Reply by PaleNeutron** [reply]
@CaselIT , I found in mapper events as above, connection has no attribute `run_async` and if I simply use `connection.execute`. It will block all async tasks. Is there any example about use async connection in mapper events?

**Reply by CaselIT** [reply]
if you use connection.execute inside an event handles it will not block the async tasks. It's documented here https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#examples-of-event-listeners-with-async-engines-sessions-sessionmakers

**Reply by PaleNeutron** [reply]
@CaselIT Here is a test, and I got quite strange result

```python
import asyncio
import datetime
import json

from sqlalchemy import (
    JSON,
    Column,
    DateTime,
    Integer,
    String,
    event,
    inspect,
    select,
    text,
)
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import declarative_base, object_session, sessionmaker

# 1. Define the Base for our ORM models
Base = declarative_base()


# 2. Define the main Portfolio model
class Portfolio(Base):
    """
    The main table holding the current state of portfolios.
    """

    __tablename__ = "portfolio"

    id = Column(Integer, primary_key=True)
    name = Column(String(100), nullable=False)
    description = Column(String(255))

    def __repr__(self):
        return f"<Portfolio(id={self.id}, name='{self.name}')>"


# 3. Define the history/audit model
class PortfolioHis(Base):
    """
    The history table to store changes made to the Portfolio table.
    """

    __tablename__ = "portfolio_his"

    id = Column(Integer, primary_key=True)
    portfolio_id = Column(Integer, nullable=False, index=True)
    changed_at = Column(DateTime, default=datetime.datetime.utcnow)

    # The 'changes' column will store a JSON object detailing what changed.
    # e.g., {'name': {'old': 'Old Name', 'new': 'New Name'}}
    changes = Column(JSON)

    def __repr__(self):
        return f"<PortfolioHis(id={self.id}, portfolio_id={self.portfolio_id})>"


# 4. The core of the solution: The event listener
@event.listens_for(Portfolio, "after_update")
def receive_after_update(mapper, connection, target):
    """
    Listen for the 'after_update' event on the Portfolio model.

    This function is called after a Portfolio object has been updated
    in a session flush, but before the transaction is committed.
    """
    # create a test sql here, sleep for 10 seconds by sql
    t = datetime.datetime.now()
    print(f"{target} updated at {t}")
    connection.execute(
        text("SELECT pg_sleep(10);")
    )  # Simulate a long-running operation
    print(f"Finished processing {target} at {datetime.datetime.now()}")


async def main():
    """test if the event listener will block whole thread by run multiple tasks"""
    engine = create_async_engine(
        "postgresql+asyncpg://postgres:password@localhost/template1",
        echo=False,  # Set to True to see the generated SQL
        pool_size=10,  # Adjust the pool size as needed
    )
    async with engine.begin() as conn:
        # Create the tables
        await conn.run_sync(Base.metadata.create_all)

    async_session = sessionmaker(
        bind=engine,
        class_=AsyncSession,
        expire_on_commit=False,
    )

    async with async_session() as session:
        # Create a new portfolio
        stmt = select(Portfolio).limit(1)
        portfolio = await session.scalar(stmt)
        if not portfolio:
            portfolio = Portfolio(
                name="Test Portfolio", description="A test portfolio"
            )
            session.add(portfolio)
            await session.commit()
        portfolio_id = portfolio.id

    # Run multiple tasks to test if the event listener blocks the thread
    async def update_portfolio(new_name):
        async with async_session() as session:
            portfolio = await session.get(Portfolio, portfolio_id)
            print(f"Updating portfolio to {new_name}")
            if portfolio:
                portfolio.name = new_name
                await session.commit()

    tasks = [
        asyncio.create_task(update_portfolio(f"Task {i} Portfolio")) for i in range(5)
    ]
    # execute all tasks concurrently without blocking

    await asyncio.sleep(1)  # Simulate other async work
    print("Tasks have been started. Now doing some other work...")
    await asyncio.sleep(1)  # Simulate other async work
    print("Finished with the other work.")
    await asyncio.gather(*tasks)
    await engine.dispose()  # Clean up the engine



if __name__ == "__main__":
    asyncio.run(main())

```

the output is:

```
Updating portfolio to Task 0 Portfolio
<Portfolio(id=1, name='Task 0 Portfolio')> updated at 2025-06-27 11:24:35.268699
Updating portfolio to Task 1 Portfolio
Updating portfolio to Task 3 Portfolio
Updating portfolio to Task 2 Portfolio
Updating portfolio to Task 4 Portfolio
<Portfolio(id=1, name='Task 1 Portfolio')> updated at 2025-06-27 11:24:35.415836
Tasks have been started. Now doing some other work...
Finished with the other work.
Finished processing <Portfolio(id=1, name='Task 0 Portfolio')> at 2025-06-27 11:24:45.271265
Finished processing <Portfolio(id=1, name='Task 1 Portfolio')> at 2025-06-27 11:24:45.417233
<Portfolio(id=1, name='Task 2 Portfolio')> updated at 2025-06-27 11:24:45.453631
Finished processing <Portfolio(id=1, name='Task 2 Portfolio')> at 2025-06-27 11:24:55.465548
<Portfolio(id=1, name='Task 4 Portfolio')> updated at 2025-06-27 11:24:55.495053
Finished processing <Portfolio(id=1, name='Task 4 Portfolio')> at 2025-06-27 11:25:05.504068
<Portfolio(id=1, name='Task 3 Portfolio')> updated at 2025-06-27 11:25:05.527979
Finished processing <Portfolio(id=1, name='Task 3 Portfolio')> at 2025-06-27 11:25:15.534339
```

`Finished with the other work.` comes before `Finished processing ` seems indicate that `connection.execute` do not block the main thread. But tasks are running in order and takes 5 * 10s. I except they finish in 10s together.

It seems there is another thread for event and executing sync query in it will not block main thread but will block other event.

Is there any way to use async connection in event listener?

**Reply by CaselIT** [reply]
If i'm not mistaken the after update event happens in the same transaction than the update.
Postgresql will likely serialize the updates, since you are always updating the same row. Create one row per task and they should finish in 10 seconds.

**Reply by PaleNeutron** [reply]
@CaselIT , Thanks, it works as excepted after I change to update different rows.

**Reply by CaselIT** [reply]
Thank for confirming!
